File: client_leger/lib/services/chat_service.dart
import 'package:client_leger/environments/environment.dart';
import 'package:client_leger/interfaces/chat/chat_create.dart';
import 'package:client_leger/interfaces/chat/chat_data.dart';
import 'package:client_leger/interfaces/chat/chat_join.dart';
import 'package:client_leger/interfaces/chat/chat_message.dart';
import 'package:client_leger/interfaces/chat/message_data.dart';
import 'package:client_leger/interfaces/chat/quit_chat.dart';
import 'package:client_leger/interfaces/chat/select_chat.dart';
import 'package:client_leger/interfaces/socket_event.dart';
import 'package:client_leger/services/interceptor.dart';
import 'package:client_leger/services/socket_service.dart';
import 'package:events_emitter/events_emitter.dart';
import 'package:flutter/foundation.dart';
import 'package:get/get.dart';
import 'package:http_interceptor/http_interceptor.dart';

class ChatService extends GetxController {
  static ChatService get to => Get.find();
  String chatName = 'alloaa';
  String activeChatId = '';
  String userId = '';

  List<String> usernames = [];
  String test = 'yo';
  final httpClient =
      InterceptedClient.build(interceptors: [AuthenticationInterceptor()]);
  final String baseUrl = environment['serverUrl'];
  final messagesNotifier = ValueNotifier<List<MessageData>>([]);
  final joinableChats = ValueNotifier<List<ChatData>>([]);
  final joinedChats = ValueNotifier<List<ChatData>>([]);
  final events = EventEmitter();

  @override
  void onInit() {
    SocketService.to.on(SocketEvent.message, _newMessageHandler);
    SocketService.to.on(SocketEvent.updatedChatsList, _updateChatList);
    SocketService.to.on(SocketEvent.selectChat, _selectChatEvent);
    SocketService.to.on(SocketEvent.getUserId, _getUserId);
    SocketService.to.on(SocketEvent.quitChat, _quitChatEvent);
    super.onInit();
  }

  @override
  void onClose() {
    SocketService.to.removeEventListener(_newMessageHandler);
    SocketService.to.removeEventListener(_updateChatList);
    SocketService.to.removeEventListener(_selectChatEvent);
    SocketService.to.removeEventListener(_getUserId);
    SocketService.to.removeEventListener(_quitChatEvent);
    super.onClose();
  }

  void _newMessageHandler(dynamic message) {
    ChatMessage chatMessage = ChatMessage.fromJson(message);
    MessageData newMessage =
        MessageData(chatMessage: chatMessage, authorId: message['authorId']);
    receiveMessage(newMessage);
  }

  _updateChatList(dynamic chatData) {
    ChatData newChat = ChatData.fromJson(chatData);
    updateChatList(newChat);
  }

  _selectChatEvent(dynamic data) {
    List<MessageData> messages = [];

    for (var i = 0; i < data.length; i++) {
      messages.add(MessageData(
          chatMessage: ChatMessage.fromJson(data[i]['chatMessage']),
          authorId: data[i]['authorId']));
    }
    messagesNotifier.value = [...messagesNotifier.value, ...messages];
  }

  _getUserId(id) => userId = id;

  _quitChatEvent(chatId) => quitChatEvent(chatId);

  Future<void> createChat(String chatName, String username) async {
    final response = await httpClient.get(Uri.parse('$baseUrl/chat/$username'));
    if (response.statusCode == 200) {
      CreateChat newChat = CreateChat(name: chatName, userId: response.body);
      SocketService.to.send(SocketEvent.createChat, newChat);
    }
  }

  Future<void> getJoinedChats(String username) async {
    List<ChatData> chats = [];
    SocketService.to.send(
        SocketEvent.getJoinedChatsCallBack,
        username,
        (data) => {
              if (data is List<dynamic>)
                {
                  chats = data.map((chat) {
                    return ChatData(
                      chatId: chat['chatId'],
                      chatname: chat['chatname'],
                      isOwner: chat['isOwner'],
                    );
                  }).toList(),
                  joinedChats.value = [...chats],
                }
            });
  }

  Future<void> getJoinableChats(String username) async {
    List<ChatData> chats = [];

    SocketService.to.send(
        SocketEvent.getJoinableChatsCallBack,
        username,
        (data) => {
              if (data is List<dynamic>)
                {
                  chats = data.map((chat) {
                    return ChatData(
                      chatId: chat['chatId'],
                      chatname: chat['chatname'],
                      isOwner: chat['isOwner'],
                    );
                  }).toList(),
                  joinableChats.value = [...chats],
                }
            });
  }

  Future<void> joinChat(String chatId, String username) async {
    JoinChat joinChat = JoinChat(chatID: chatId, user: username);
    SocketService.to.send(SocketEvent.joinChat, joinChat, (response) {});
    selectChat(chatId, username);
  }

  selectChat(String chatId, String username) {
    print('selecting chat');
    activeChatId = chatId;
    messagesNotifier.value = [];
    SelectChat selectChat = SelectChat(id: chatId, user: username);
    SocketService.to.send(SocketEvent.selectChat, selectChat);
  }

  receiveMessage(MessageData message) {
    if (message.chatMessage.destination == activeChatId) {
      messagesNotifier.value = [...messagesNotifier.value, message];
    }
  }

  sendMessage(ChatMessage message) {
    if (message.content.trim().isEmpty || message.destination.length == 0) {
      return;
    }
    this.activeChatId = message.destination;

    SocketService.to.send(SocketEvent.message, message);
  }

  getUserId(String username) {
    print('getting user id');
    SocketService.to.send(SocketEvent.getUserId, username);
  }

  Future<String> getChatName(String chatId) async {
    var response =
        await httpClient.get(Uri.parse('$baseUrl/chat/chatname/$chatId'));
    chatName = response.body;
    return response.body;
  }

  void quitChat(String chatId, String username) {
    // FIX ME
    //joinableChats.removeWhere((chat) => chat.chatId == chatId);
    // joinedChats.removeWhere((chat) => chat.chatId == chatId);

    QuitChat quitChat = QuitChat(chatID: chatId, user: username);
    SocketService.to.send(SocketEvent.quitChat, quitChat);
  }

  quitChatEvent(String chatId) {
    // FIX ME
    //joinableChats.removeWhere((chat) => chat.chatId == chatId);
    // joinedChats.removeWhere((chat) => chat.chatId == chatId);
    if (activeChatId == chatId) {
      quitActiveChatEvent();
      activeChatId = '';
    }
  }

  updateChatList(ChatData newChat) {
    if (newChat.isOwner) {
      joinedChats.value = [...joinedChats.value, newChat];
    } else {
      joinableChats.value = [...joinableChats.value, newChat];
    }
    update();
  }

  quitActiveChatEvent() {
    events.emit('quit');
  }

  getEvents() {
    return events;
  }

  resetMessages() {
    print('resetting messages');
    messagesNotifier.value = [];
  }
}

File: client/src/app/services/chat/chat.service.ts
/* eslint-disable @typescript-eslint/no-empty-function */
/* eslint-disable no-unused-vars */
/* eslint-disable @typescript-eslint/no-unused-vars */
import { HttpClient } from '@angular/common/http';
import { EventEmitter, Injectable, OnDestroy } from '@angular/core';
import { SocketService } from '@app/services/socket/socket.service';
import { ChatData, MessageData, UsersOfChat } from '@common/interfaces/chat/chat.interface';
import { ChatSessionEvent } from '@common/model/events/chat-session.events';
import { ChatMessage } from '@common/model/message';
import { Observable, Subscription } from 'rxjs';
import { environment } from 'src/environments/environment';

@Injectable()
export class ChatService implements OnDestroy {
    quitActiveChat: EventEmitter<void> = new EventEmitter<void>();

    messages: MessageData[] = [];
    activeChatId: string = '';
    userId: string = '';
    joinedChats: ChatData[] = [];
    joinableChats: ChatData[] = [];
    usernames: string[] = [];
    private socketCreatedSubscription: Subscription;
    private readonly baseUrl: string = environment.serverUrl;

    constructor(private socketService: SocketService, private readonly http: HttpClient) {
        this.socketCreatedSubscription = this.socketService.executeOnSocketCreation({
            next: (isSocketCreated: boolean) => {
                if (isSocketCreated) {
                    this.setSocketEvents();
                }
            },
        });
    }

    ngOnDestroy(): void {
        this.messages = [];
        this.socketCreatedSubscription.unsubscribe();
    }

    setSocketEvents() {
        this.socketService.on(ChatSessionEvent.Message, (message: MessageData) => {
            this.receiveMessage(message);
        });

        this.socketService.on(ChatSessionEvent.UpdatedChatsList, (newChat: ChatData) => {
            this.updateChatList(newChat);
        });

        this.socketService.on(ChatSessionEvent.GetJoinableChats, (chats: ChatData[]) => {
            this.joinableChats = chats;
        });

        this.socketService.on(ChatSessionEvent.GetJoinedChats, (chats: ChatData[]) => {
            this.joinedChats = chats;
        });

        this.socketService.on(ChatSessionEvent.SelectChat, (chat: MessageData[]) => {
            this.selectChatEvent(chat);
        });

        this.socketService.on(ChatSessionEvent.GetUserId, (userId: string) => {
            this.userId = userId;
        });

        this.socketService.on(ChatSessionEvent.UsersOfChat, (request: UsersOfChat) => {
            if (this.activeChatId === request.chatID) this.usernames = request.users;
        });

        this.socketService.on(ChatSessionEvent.QuitChat, (request: { chatID: string }) => {
            this.quitChatEvent(request.chatID);
        });
    }

    removeSocketEvents() {
        this.socketService.off(ChatSessionEvent.Message, (message: MessageData) => {});
        this.socketService.off(ChatSessionEvent.UpdatedChatsList, (newChat: ChatData) => {});
        this.socketService.off(ChatSessionEvent.GetJoinableChats, (chats: ChatData[]) => {});
        this.socketService.off(ChatSessionEvent.GetJoinedChats, (chats: ChatData[]) => {});
        this.socketService.off(ChatSessionEvent.SelectChat, (chat: MessageData[]) => {});
        this.socketService.off(ChatSessionEvent.GetUserId, (userId: string) => {});
        this.socketService.off(ChatSessionEvent.UsersOfChat, (request: UsersOfChat) => {});
        this.socketService.off(ChatSessionEvent.QuitChat, (request: { chatID: string }) => {});
    }

    createChat(name: string, userId: string) {
        const newChat = { name, userId };
        this.socketService.send(ChatSessionEvent.CreateChat, newChat);
    }

    getJoinableChats(username: string) {
        this.socketService.send(ChatSessionEvent.GetJoinableChats, username);
    }

    getJoinedChats(username: string) {
        this.socketService.send(ChatSessionEvent.GetJoinedChats, username);
    }

    joinChat(chatId: string, username: string) {
        this.socketService.send(ChatSessionEvent.JoinChat, { chatID: chatId, user: username }, () => {
            this.selectChat(chatId, username);
        });
    }

    selectChat(chatId: string, username: string) {
        this.messages = [];
        this.activeChatId = chatId;
        this.joinableChats = this.joinableChats.filter((chat) => chat.chatId !== chatId);
        this.socketService.send(ChatSessionEvent.SelectChat, { id: chatId, user: username });
    }

    receiveMessage(message: { chatMessage: ChatMessage; authorId: string }) {
        if (this.activeChatId !== message.chatMessage.destination) {
            return;
        }
        this.messages.push(message);
    }

    sendMessage(message: ChatMessage) {
        if (message.content.length < 1 || message.destination.length < 1) {
            return;
        }
        this.activeChatId = message.destination;
        this.socketService.send(ChatSessionEvent.Message, message);
    }

    getUserId(username: string) {
        this.socketService.send(ChatSessionEvent.GetUserId, username);
    }

    lookUpUser(username: string): Observable<string> {
        return this.http.get(`${this.baseUrl}/chat/${username}`, { responseType: 'text' });
    }

    getChatName(chatId: string): Observable<string> {
        return this.http.get(`${this.baseUrl}/chat/chatname/${chatId}`, { responseType: 'text' });
    }

    quitChat(chatId: string, username: string) {
        this.joinedChats = this.joinedChats.filter((chat) => chat.chatId !== chatId);
        this.socketService.send(ChatSessionEvent.QuitChat, { chatID: chatId, user: username });
    }

    quitChatEvent(chatID: string) {
        this.joinedChats = this.joinedChats.filter((chat) => chat.chatId !== chatID);
        this.joinableChats = this.joinableChats.filter((chat) => chat.chatId !== chatID);
        if (this.activeChatId === chatID) {
            this.quitActiveChatEvent();
            this.activeChatId = '';
        }
    }

    selectChatEvent(chat: MessageData[]) {
        if (!chat.length) {
            this.messages = [];
            return;
        }
        this.messages = chat;
    }

    updateChatList(newChat: ChatData) {
        if (newChat.isOwner) {
            this.joinedChats.push(newChat);
        } else {
            this.joinableChats.push(newChat);
        }
    }

    quitActiveChatEvent() {
        this.quitActiveChat.emit();
    }

    getQuitActiveChatEvent(): EventEmitter<void> {
        return this.quitActiveChat;
    }

    getUsersOfChat(chatId: string): Observable<string[]> {
        return this.http.get<string[]>(`${this.baseUrl}/chat/users/${chatId}`);
    }

    receiveMessageReplay(message: ChatMessage) {
        this.messages.push({ chatMessage: message, authorId: message.author });
    }
}
